### PR TITLE
OCPBUGS-11885: use bond members as keys instead of bond interface

### PR DESCRIPTION
**Pull request by msherif1234**
with running rhel9 we have now bond XDP native support by design xdp packets will be carrying member interface index not the bond interface, while we still only
attach to bond interface

**- What this PR does and why is it needed**
Fix XDP on bond interface with rhel9 where it brings in native XDP driver

**- Special notes for reviewers**

**- How to verify it**
- bring up OCP cluster with the changes
- verify keys have the member interface index instead of the bond

```bash
5: packet0: <BROADCAST,MULTICAST,SLAVE,UP,LOWER_UP> mtu 1500 qdisc mq master bond0 state UP mode DEFAULT group default qlen 1000
    link/ether 50:7c:6f:13:cc:92 brd ff:ff:ff:ff:ff:ff
    altname enp1s0f0
6: packet1: <BROADCAST,MULTICAST,SLAVE,UP,LOWER_UP> mtu 1500 qdisc mq master bond0 state UP mode DEFAULT group default qlen 1000
    link/ether 50:7c:6f:13:cc:92 brd ff:ff:ff:ff:ff:ff permaddr 50:7c:6f:13:cc:93
    altname enp1s0f1
7: bond0: <BROADCAST,MULTICAST,MASTER,UP,LOWER_UP> mtu 1500 xdp qdisc noqueue master ovs-system state UP mode DEFAULT group default qlen 1000
    link/ether 50:7c:6f:13:cc:92 brd ff:ff:ff:ff:ff:ff
    prog/xdp id 4 name ingress_node_fi tag c35fd8443c61125e jited 
```

- daemon logs
```bash
2023/05/08 19:30:05 Adding or updating ingress firewall rules for key {64 5 [147 75 61 7 0 0 0 0 0 0 0 0 0 0 0 0]}
2023/05/08 19:30:05 Adding or updating ingress firewall rules for key {64 6 [147 75 61 7 0 0 0 0 0 0 0 0 0 0 0 0]}
2023/05/08 19:30:05 Adding or updating ingress firewall rules for key {96 5 [252 0 248 83 12 205 231 147 0 0 0 0 0 0 0 0]}
2023/05/08 19:30:05 Adding or updating ingress firewall rules for key {96 6 [252 0 248 83 12 205 231 147 0 0 0 0 0 0 0 0]}
```

- bpftool map dump
```bash
[root@worker-00 /]# bpftool  map dump id 18 | grep -F "147,75,61,7" -A 10 -B 10
[{
        "key": {
            "prefixLen": 64,
            "ingress_ifindex": 5,
            "ip_data": [147,75,61,7,0,0,0,0,0,0,0,0,0,0,0,0
            ]
        },
        "value": {
            "rules": [{
                    "ruleId": 0,
                    "protocol": 0,
                    "dstPortStart": 0,
                    "dstPortEnd": 0,
                    "icmpType": 0,
                    "icmpCode": 0,
--
                    "icmpType": 0,
                    "icmpCode": 0,
                    "action": 0
                }
            ]
        }
    },{
        "key": {
            "prefixLen": 64,
            "ingress_ifindex": 6,
            "ip_data": [147,75,61,7,0,0,0,0,0,0,0,0,0,0,0,0
            ]
        },
        "value": {
            "rules": [{
                    "ruleId": 0,
                    "protocol": 0,
                    "dstPortStart": 0,
                    "dstPortEnd": 0,
                    "icmpType": 0,
                    "icmpCode": 0,
```

- run curl traffic from work01 and check events container
```bash
Starting pod/worker-01temp-test07qedevclusteropenshiftcom-debug ...
To use host binaries, run `chroot /host`
Pod IP: 147.75.61.7
If you don't see a command prompt, try pressing enter.
sh-4.4# curl 145.40.89.17:8080
```

- events
```bash
]$ oc logs -n openshift-ingress-node-firewall ingress-node-firewall-daemon-vv89g  -c events --follow
2023-05-08 19:34:53 +0000 UTC worker-00.temp-test07.qe.devcluster.openshift.com ruleId 20 action Drop len 74 if packet0
2023-05-08 19:34:53 +0000 UTC worker-00.temp-test07.qe.devcluster.openshift.com         ipv4 src addr 147.75.61.7 dst addr 145.40.89.17
2023-05-08 19:34:53 +0000 UTC worker-00.temp-test07.qe.devcluster.openshift.com         tcp srcPort 36036 dstPort 8080
2023-05-08 19:34:54 +0000 UTC worker-00.temp-test07.qe.devcluster.openshift.com ruleId 20 action Drop len 74 if packet0
2023-05-08 19:34:54 +0000 UTC worker-00.temp-test07.qe.devcluster.openshift.com         ipv4 src addr 147.75.61.7 dst addr 145.40.89.17
2023-05-08 19:34:54 +0000 UTC worker-00.temp-test07.qe.devcluster.openshift.com         tcp srcPort 36036 dstPort 8080
2023-05-08 19:34:56 +0000 UTC worker-00.temp-test07.qe.devcluster.openshift.com ruleId 20 action Drop len 74 if packet0
2023-05-08 19:34:56 +0000 UTC worker-00.temp-test07.qe.devcluster.openshift.com         ipv4 src addr 147.75.61.7 dst addr 145.40.89.17
2023-05-08 19:34:56 +0000 UTC worker-00.temp-test07.qe.devcluster.openshift.com         tcp srcPort 36036 dstPort 8080
2023-05-08 19:35:00 +0000 UTC worker-00.temp-test07.qe.devcluster.openshift.com ruleId 20 action Drop len 74 if packet0
2023-05-08 19:35:00 +0000 UTC worker-00.temp-test07.qe.devcluster.openshift.com         ipv4 src addr 147.75.61.7 dst addr 145.40.89.17
2023-05-08 19:35:00 +0000 UTC worker-00.temp-test07.qe.devcluster.openshift.com         tcp srcPort 36036 dstPort 8080
2023-05-08 19:35:08 +0000 UTC worker-00.temp-test07.qe.devcluster.openshift.com ruleId 20 action Drop len 74 if packet0
2023-05-08 19:35:08 +0000 UTC worker-00.temp-test07.qe.devcluster.openshift.com         ipv4 src addr 147.75.61.7 dst addr 145.40.89.17

```

- unit-test
```bash
[mmahmoud@localhost interfaces]$ go test -exec sudo ./... -v
=== RUN   TestIsValidInterfaceNameAndState
--- PASS: TestIsValidInterfaceNameAndState (0.00s)
=== RUN   TestGetInterfaceIndices
    interfaces_test.go:78: By creating bond interfaces with members
    interfaces_test.go:83: bond interface index 13
    interfaces_test.go:111: packet1 interface index 14
    interfaces_test.go:111: packet2 interface index 15
    interfaces_test.go:111: packet3 interface index 16
    interfaces_test.go:111: packet4 interface index 17
--- PASS: TestGetInterfaceIndices (0.11s)
PASS
ok      github.com/openshift/ingress-node-firewall/pkg/interfaces       0.133s
```
**- Description for the changelog**
Fix Ingress node firewall on bond interface running rhel9

### CODE REVIEW

File: pkg/interfaces/interfaces.go [hunk 1]
```diff
@@ -78,3 +78,39 @@ func DetachXDPFromAllInterfaces() error {
...
+// will be using bond member's interface_indices not the bond interface_index.
+func GetInterfaceIndices(interfaceName string) ([]uint32, error) {
+	var membersList []uint32
+
+	link, err := netlink.LinkByName(interfaceName)
+	if err != nil {
+		return nil, err
+	}
+
+	if link.Type() != "bond" {
```
Comment (danwinship) [inline pkg/interfaces/interfaces.go:93]: Are we sure the change in behavior only applies to bond interfaces and not any other types? And that this is the best way to identify bond interfaces?

Also, will this change mean that now the latest version of infw doesn't work correctly with bonds on RHEL8?

Should we add rules for both the member ifIDs and the master ifID?

Comment (msherif1234) [reply pkg/interfaces/interfaces.go:93]: the latest won't work with rhel8 for other things as well like enabling jumpo supporyt/apis, we can't enable XDP on members and bond we will get error already exists

**Review by dave-tucker — APPROVED** [review]
As far as I understand the issue, this code change LGTM.
Whether or not XDP support for bonded interfaces in RHEL makes any sense or not is another issue.
I would recommend we file a bug with the kernel team to get this behaviour changed - i.e revert XDP bond support completely + manually attach the XDP prog to members of the bond.

**Pull request merged**